src/pcp/gadgets.rs

```diff
@@ -214,7 +221,271 @@ impl<F: FieldElement> Gadget<F> for PolyEval<F> {
+        let mut z_len = y.len();
+        z[..y.len()].clone_from_slice(y);
+
+        for i in 0..self.poly.len() {
+            for j in 0..z_len {
+                outp[j] += self.poly[i] * z[j];
+            }
+
+            if i < self.poly.len() - 1 {
+                for j in 0..n {
+                    z_vals[j] *= x_vals[j];
+                }
+
+                discrete_fourier_transform(&mut z, &z_vals, n)?;
+                discrete_fourier_transform_inv_finish(&mut z, n, self.n_inv);
+                z_len += x.len();
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<F: FieldElement> Gadget<F> for BlindPolyEval<F> {
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        gadget_call_check(self, inp.len())?;
+        Ok(inp[1] * poly_eval(&self.poly, inp[0]))
+    }
+
+    fn call_poly(&mut self, outp: &mut [F], inp: &[Vec<F>]) -> Result<(), PcpError> {
+        gadget_call_poly_check(self, outp, inp)?;
+
+        for x in outp.iter_mut() {
+            *x = F::zero();
+        }
+
+        if inp[0].len() >= FFT_THRESHOLD {
+            self.call_poly_fft(outp, inp)
+        } else {
+            self.call_poly_direct(outp, inp)
+        }
+    }
+
+    fn arity(&self) -> usize {
+        2
+    }
+
+    fn degree(&self) -> usize {
+        poly_deg(&self.poly) + 1
+    }
+
+    fn calls(&self) -> usize {
+        self.num_calls
+    }
+
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+}
+
+/// Marker trait for abstracting over [`SerialSum`] and [`ParallelSum`]
+pub trait SumGadget<F: FieldElement, G>: Gadget<F> + Debug {
+    /// Wraps `inner` into a sum gadget with `chunks` chunks
+    fn new(inner: G, chunks: usize) -> Self;
+}
+
+/// A wrapper gadget that applies the inner gadget to chunks of input and returns the sum of the
+/// outputs. The arity is equal to the arity of the inner gadget times the number of chunks.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SerialSum<F: FieldElement, G: Gadget<F>> {
+    inner: G,
+    chunks: usize,
+    phantom: PhantomData<F>,
+}
+
+impl<F: FieldElement, G: 'static + Gadget<F>> SumGadget<F, G> for SerialSum<F, G> {
+    fn new(inner: G, chunks: usize) -> Self {
+        Self {
+            inner,
+            chunks,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<F: FieldElement, G: 'static + Gadget<F>> Gadget<F> for SerialSum<F, G> {
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        gadget_call_check(self, inp.len())?;
+        let mut outp = F::zero();
+        for chunk in inp.chunks(self.inner.arity()) {
+            outp += self.inner.call(chunk)?;
+        }
+        Ok(outp)
+    }
+
+    fn call_poly(&mut self, outp: &mut [F], inp: &[Vec<F>]) -> Result<(), PcpError> {
+        gadget_call_poly_check(self, outp, inp)?;
+
+        let res = inp
+            .chunks(self.inner.arity())
+            .map(|chunk| {
+                let mut partial_outp = vec![F::zero(); outp.len()];
```

Comment:
Per discussion I rewrote this as a `for` loop instead of using `.map().reduce()`, which lets us allocate `partial_outp` once and avoid awkward closure capture rules.